Implement social sharing

sdotson opened the pull request:
This is tackling #73.

It is currently using a dummy app id that I created, but will be updated with the official one once that is created. I'm not entirely sure about my approach, so I'm absolutely open and humble to feedback!

Specific things I wasn't sure about:
- The best way to include the fb sdk
- Changing the Component from 'dumb' to a class
- The way to implement the click handler that wouldn't trigger lint errors
- Anything else I might have forgotten

One final note is that the facebook share dialogue won't work with localhost links.

dorono commented on src/components/SocialBtns/SocialBtns.js:
```diff
@@ -1,30 +1,50 @@
-  );
-};
+class SocialBtns extends Component {
+
```
Whenever you add a class, if you were to be binding functions in the class, you would add as the first method in the class:
```
constructor(props){
     super(props);
}
```
See https://facebook.github.io/react/docs/react-component.html#constructor for more details

Otherwise, your syntax looks good!

sdotson replied:
I had trouble getting your suggestion to work. It looks like the `constructor` method doesn't have access to to `fbShareHandler` through `this`. Perhaps I missed something.

dorono commented on src/components/SocialBtns/SocialBtns.js:
```diff
@@ -1,30 +1,50 @@
+        >
+        </a>
+        <a href={twitterLink} className={styles.twitterBtn}>
+          <FaTwitter size={iconSize} />
```
I unfortunately don't have time to pull down and run your PR, but I was thinking it would be ideal if this link opened up a modal window for sharing, or at the very least, opened in a new browser tab, just to make sure that the user isn't taken away from our site.

sdotson replied:
The facebook sdk code with the onClick handler opens up a little window directing the user to share on facebook.

dorono commented on src/components/SocialBtns/SocialBtns.js:
```diff
@@ -1,30 +1,50 @@
-import React, { PropTypes } from 'react';
+/* global FB, location */
```
What's this comment for?

sdotson replied:
Yep. This is so the linter ignores the global variables FB and location.

pfarnach replied:
It may be a little clearer to use `window.location` (and `window` is already globally ignored in our `.eslintrc`).  

And why is `FB` a global/where is that coming from?  I'd prefer if we used imports where possible, or if something is attaching it to `window`, we use `window.FB` and make sure it's well commented

dorono commented:
Looks good! Like I said, I didn't get a chance to pull it down, but if you make those fixes and it works, we should probably be good, although I'd be curious to see if some of the more React-savvy folks here might have more to say, since I'm relatively new to React myself. Any thoughts, @pdw207  or @pfarnach? 

Another idea that might simplify things...perhaps there is an existing, light-weight 3rd party social sharing open source React component that we could use in case it makes things easier in terms of the user experience, using on localhost, etc. Always good to avoid those if possible, but maybe an idea if we run into problems. :)

pfarnach commented:
@sdotson It seems like a test is broken too

pfarnach commented on src/components/SocialBtns/SocialBtns.test.js:
```diff
@@ -1,26 +1,13 @@
   it('renders without crashing', () => {
     const wrapper = shallow(<SocialBtns {...props} />);
     expect(wrapper).toHaveLength(1);
   });
```
It seems like these tests should be updated, not removed

pfarnach commented on src/components/SocialBtns/fb_sdk.js:
```diff
@@ -0,0 +1,21 @@
+  init: function() {
+    window.fbAsyncInit = function() {
+      FB.init({
+        appId      : '339880856413925',
```
I'm not familiar with implementing sharing so this is new territory for me.  A couple thoughts/questions:
1.) Is this `appId` secret / do we want it committed to version control?
2.) We're using ES2015 imports in the frontend code, could you switch this over?
3.) Our file-naming convention is to camel case everything, so this would be something like `fbSdk.js` or `fbSDK.js`

pdw207 replied:
This follows [facebook quick start instructions](https://developers.facebook.com/docs/javascript/quickstart). The appID is a unique identifier for the app. We also have an app secret that can't be shared. With that said should maybe we should create a separate config file with all keys and ids of external resources . We could also pull out the email used for trello in webpack/parts.json and put it in config.

sdotson commented:
Thanks everybody for reviewing. I certainly learned a few things. 

Oh Lord, do I feel silly. There's no need for the facebook sdk for a social share link. There's a simpler way with `https://www.facebook.com/sharer/sharer.php?u=`.

I just now reverted most changes, implemented a much simpler url approach, and made sure both twitter and facebook links opened in a new window.

pfarnach commented on src/components/SocialBtns/SocialBtns.js:
```diff
@@ -1,24 +1,32 @@
+/* global FB, location */
```
It seems like you can use `window.location` and get rid of this line

sdotson replied:
Good suggestion. Done.

pdw207 commented on src/components/SocialBtns/SocialBtns.js:
```diff
@@ -1,24 +1,32 @@
 const SocialBtns = ({
-  fbLink = 'https://www.facebook.com',
-  twitterLink = 'https://twitter.com',
+  fbLink = `https://www.facebook.com/sharer/sharer.php?u=${location.href}`,
```
<img width="859" alt="screen shot 2017-04-10 at 12 08 16 pm" src="https://cloud.githubusercontent.com/assets/4714249/24871184/cfebfaca-1de6-11e7-803d-b7a1fdd6aeca.png">

Can you include meta tags in the head to include title, image and description?

```
 <meta property="og:url"           content="http://www.your-domain.com/your-page.html" />
  <meta property="og:type"          content="website" />
  <meta property="og:title"         content="Your Website Title" />
  <meta property="og:description"   content="Your description" />
  <meta property="og:image"         content="http://www.your-domain.com/path/image.jpg" />
```

pdw207 replied:
Maybe concat the title with date as well.

sdotson replied:
It sounds like the only way to get the Facebook parser to recognize dynamically set og meta tags might be to render on the server ([based on this](https://github.com/nfl/react-helmet/issues/213)). I admittedly don't know a whole lot about server-rendered React. Is that the current set up?

pdw207 commented on src/components/SocialBtns/SocialBtns.js:
```diff
@@ -1,24 +1,32 @@
-  fbLink = 'https://www.facebook.com',
-  twitterLink = 'https://twitter.com',
+  fbLink = `https://www.facebook.com/sharer/sharer.php?u=${location.href}`,
+  twitterLink = `https://twitter.com/intent/tweet?text=${location.href}`,
```
Can we include the image, title and date? Right now its just a link that will be shortened by twitter's service with no other contextual information.

sdotson replied:
Looks like Twitter also uses the same  [open graph (og) meta tags as facebook, along with some of their own](https://dev.twitter.com/cards/markup).

sdotson commented:
What do you all think about merging what we have here and creating a new issue to make this a server-rendered universal react app so we can do the dynamic meta tags for facebook/twitter share? Not sure how much work that would entail but I would find out after some googlin'/stackoverflowin'.

dorono commented:
LOL, funny you mention that...initially it was server-side-rendered, but a lot of us are new to server-side rendering and got stuck on tasks that we would have been able to complete if not for the server-side rendering learning curve. So we went back to client-side rendering. 

Anyhow, my vote is not to spend too much more time on this, we can always refine it later. 

@sdotson If you don't mind posting a screenshot of how the popups look, as long as they look OK and the unit tests are passing, then I say we merge it and refine as needed in future iterations?

sdotson commented:
@dorono sure thing.

<img width="844" alt="screen shot 2017-04-14 at 10 38 32 am" src="https://cloud.githubusercontent.com/assets/5664775/25046396/3e2217b6-20ff-11e7-83ab-3dde25eeef3a.png">
<img width="615" alt="screen shot 2017-04-14 at 10 40 20 am" src="https://cloud.githubusercontent.com/assets/5664775/25046397/3e240f44-20ff-11e7-9bb4-4a78c7c01c5b.png">

pdw207 commented:
Arash mentioned open graph: `https://developers.facebook.com/docs/opengraph/getting-started`. Looks like you started with that approach.
